### PR TITLE
A multi-GPU distributed trainer

### PR DESCRIPTION
**Pull request by mallamanis**
This PR adapts the `ptgnn` library for a multi-GPU training and contains an example for the `typilus` GNN.

#### Open Issues

- [x] Gradient clipping hangs training and all subsequent training (with or without clipping) also hang, even after a VM reboot.
- [x] Verify that results are similar to the non-distributed version.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,16 +1,15 @@
...
 
 RUN python3.8 -m pip install --upgrade wheel pip cffi
-RUN python3.8 -m pip install --no-cache-dir sentencepiece==0.1.90
-RUN python3.8 -m pip install --no-cache-dir azureml-sdk chardet datasketch docopt jedi libcst msgpack tqdm typing_extensions dpu-utils
+RUN python3.8 -m pip install --no-cache-dir azureml-sdk chardet datasketch docopt msgpack tqdm typing_extensions dpu-utils
 
 
 # ReInstall torch scatter
-RUN python3.8 -m pip install --no-cache-dir --upgrade torch-scatter -f https://pytorch-geometric.com/whl/torch-1.7.0+cu101.html
+RUN python3.8 -m pip install --no-cache-dir --upgrade torch-scatter -f https://pytorch-geometric.com/whl/torch-1.9.0+cu102.html
```
Comment (mmjb) [inline Dockerfile:12]: I believe this will break on the next torch release - would it make sense to use `pip install torch~=1.9.0` (see https://www.python.org/dev/peps/pep-0440/#compatible-release) instead?

File: ptgnn/baseneuralmodel/distributedtrainer.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+        validate_on_start: bool = True,
+        patience: int = 5,
+        initialize_metadata: bool = True,
+        parallelize: bool = True,
+        shuffle_training_data: bool = True,
+    ) -> None:
+        """
+        The training-validation loop for `AbstractNeuralModel`s.
+
+        :param training_data: An iterable that each iteration yields the full training data. Note
```
Comment (mmjb) [inline ptgnn/baseneuralmodel/distributedtrainer.py]: I'm not sure I understand this comment. Is the contract that the set of examples by iterating over the iterator on all nodes should be the full training set, or should each node do the full set?

**Review by mmjb — APPROVED** [review]
This looks reasonable, though I have to admit that I'm not very familiar with the distributed training package in torch, and so may have overlooked subtle problems...

File: Dockerfile (hunk 1 above)
Comment (mallamanis) [reply Dockerfile:12]: Thanks, will be fixed in the next commit.

File: ptgnn/baseneuralmodel/distributedtrainer.py (hunk 2 above)
Comment (mallamanis) [reply ptgnn/baseneuralmodel/distributedtrainer.py]: (updated)
